### PR TITLE
ST_ClusterIntersecting / ST_ClusterWithin

### PR DESCRIPTION
**Pull request by dbaston**
Initial implementation of clustering functions described at http://trac.osgeo.org/postgis/wiki/DevClusteringFunctions

Discussion at http://trac.osgeo.org/postgis/ticket/3023

### CODE REVIEW

File: postgis/postgis.sql.in [hunk 1]
```diff
@@ -3448,6 +3448,12 @@ CREATE OR REPLACE FUNCTION pgis_geometry_polygonize_finalfn(pgis_abs)
 	AS 'MODULE_PATHNAME'
 	LANGUAGE 'c';
 
+-- Availability: 2.0
```
Comment (mwtoews) [inline postgis/postgis.sql.in]: Should be 2.2

**Comment by strk** [comment]
Why do the function return an array of geometry collections rather than single geometry collections ?

**Comment by dbaston** [comment]
@strk Because each cluster is itself a GeometryCollection, and there's no way (that I know of) to unpack only one level of a GeometryCollection.

**Comment by strk** [comment]
Had you considered returning a SETOF geometry ?

**Comment by dbaston** [comment]
@strk Considered, but I wasn't sure if an aggregate could also be a set-returning function, and in any case that seemed like an unusual semantic.  A geometry array is unusual too though, so I'm not opposed to it, except that `unnest` is already implemented and works quite nicely.

Here's an example of one way I use these functions, if that's helpful:

    WITH clusters AS
    (SELECT category, unnest(ST_ClusterWithin(geom, 1000.0/111111)) AS geom    
    FROM public.points GROUP BY category)
    
    SELECT
      row_number() over() AS cluster_id,
      category,
      ST_ConvexHull(geom) AS geom,
      ST_CollectionExtract(geom, 1) AS geom_points 
    FROM clusters
    WHERE geometrytype(ST_ConvexHull(geom)) = 'POLYGON';

**Comment by strk** [comment]
Agreed, it makes sense to return an array. Any thought about being able to include identifiers into the output, to match back to original rows ?

**Comment by dbaston** [comment]
@strk I wasn't able to think of a way to introduce identifiers without adding a lot of extra complexity.  I guess you'd essentially need to return an array of a "Feature" data type with "id" and "geom" fields.  Is "id" then restricted to a single data type, or can any Postgres type be an identifier?  Do you need new functions to extract the parts of the "Feature" data type that you're interested in?  It just seems to snowball a bit, to provide a (useful) behavior that basic Postgres aggregates (like `max`) don't offer.

Ultimately, these functions don't modify the input geometries at all, so you could join back to your input data on `ST_Equals`, or if that's too slow, maybe you could abuse the `M` or `Z` coordinates to pass identifiers through the aggregate.

**Comment by strk** [comment]
Primay key can be composed by 1 or more fields of arbitrary types, so the only way to expose it would be by encoding it to a generic format able to express them all (text[]? json?) and then return something like ``ROW(pkey text[][], cluster geometry[])``. Would make things more complex, indeed.

Abusing ``M`` or ``Z`` (or even ``SRID``) could be effective, as long as you could bend your primary key to a float8 (or a [short?] int), but not for other cases (text pkeys).

File: postgis/lwgeom_geos.c [hunk 2]
```diff
@@ -3256,6 +3263,142 @@ POSTGIS2GEOS(GSERIALIZED *pglwgeom)
...
+        }
+
+		*is3d = *is3d || gserialized_has_z(geom);
+
+		geos_geoms[i] = (GEOSGeometry*) POSTGIS2GEOS(geom);
+		if (!geos_geoms[i])   /* exception thrown at construction */
+		{
+            lwpgerror("Geometry could not be converted to GEOS");
+
+            while (i >= 0)
```
Comment (pramsey) [inline postgis/lwgeom_geos.c]: This looks like an infinite loop to me... 'i' is declared unsigned (also, Don't see it initialized to anything other than 0)

File: postgis/lwgeom_geos.c [hunk 3]
```diff
@@ -3256,6 +3263,142 @@ POSTGIS2GEOS(GSERIALIZED *pglwgeom)
...
+        GSERIALIZED *geom = (GSERIALIZED*) DatumGetPointer(value);
+
+        if (isnull)
+        {
+            continue;
+        }
+
+		*is3d = *is3d || gserialized_has_z(geom);
+
+		geos_geoms[i] = (GEOSGeometry*) POSTGIS2GEOS(geom);
```
Comment (pramsey) [inline postgis/lwgeom_geos.c]: Maybe you mean geos_geoms[i++] here?

File: postgis/lwgeom_geos.c [hunk 4]
```diff
@@ -3256,6 +3263,142 @@ POSTGIS2GEOS(GSERIALIZED *pglwgeom)
...
+		if (!gotsrid)
+		{
+			*srid = gserialized_get_srid(geom);
+            gotsrid = true;
+		}
+		else if (*srid != gserialized_get_srid(geom))
+		{
+            error_if_srid_mismatch(*srid, gserialized_get_srid(geom));
+
+            while (i >= 0)
```
Comment (pramsey) [inline postgis/lwgeom_geos.c]: See above...

File: liblwgeom/cunit/cu_unionfind.c [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+{
+	UNIONFIND *uf = UF_create(10);
+	uint32_t final_clusters[] = { 0, 2, 2, 2, 4, 5, 6, 0, 0, 2 };
+	uint32_t final_sizes[]    = { 3, 0, 4, 0, 1, 1, 1, 0, 0, 0 };
+	uf->clusters = final_clusters;
+	uf->cluster_sizes = final_sizes;
+	uf->num_clusters = 5;
+
+	uint32_t* actual_ids_by_cluster = UF_ordered_by_cluster(uf);
+	uint32_t expected_ids_by_cluster[] = { 0, 7, 8, 1, 2, 3, 9, 4, 5, 6 };
```
Comment (pramsey) [inline liblwgeom/cunit/cu_unionfind.c]: On OSX, I'm getting a result of {7,0,8,9,3,2,1,4,5,6}

**Comment by dbaston** [comment]
@pramsey interesting find in cu_unionfind.c.  I can either make `UF_ordered_by_cluster` completely deterministic, or I can make the test handle the multiple possible valid orderings.  thoughts?

**Comment by pramsey** [comment]
Well, you'll have a hard time writing online regressions for other systems if your orderings are non-deterministic, which means you'll just not test them, probably. Which isn't to say we don't have this kind of thing: a lot of the clipping code returns nondeterministic orderings... maybe just fix the test.

File: liblwgeom/cunit/cu_unionfind.c [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+	UNIONFIND *uf = UF_create(10);
+
+	uint32_t expected_initial_ids[] =   { 0, 1, 2, 3, 4, 5, 6, 7, 8, 9 };
+	uint32_t expected_initial_sizes[] = { 1, 1, 1, 1, 1, 1, 1, 1, 1, 1 };
+
+	CU_ASSERT_EQUAL(10, uf->N);
+	CU_ASSERT_EQUAL(10, uf->num_clusters);
+	CU_ASSERT_EQUAL(0, memcmp(uf->clusters, expected_initial_ids, 10*sizeof(uint32_t)));
+	CU_ASSERT_EQUAL(0, memcmp(uf->cluster_sizes, expected_initial_sizes, 10*sizeof(uint32_t)));
+}
```
Comment (pramsey) [inline liblwgeom/cunit/cu_unionfind.c]: Put a UF_destroy(uf); here for memory cleanliness...

File: liblwgeom/cunit/cu_unionfind.c [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+	UF_union(uf, 1, 2); /* add 1 (smaller) to 2-3 (larger)    */
+
+	uint32_t expected_final_ids[] =   { 0, 2, 2, 2, 4, 5, 6, 0, 0, 9 };
+	uint32_t expected_final_sizes[] = { 3, 0, 3, 0, 1, 1, 1, 0, 0, 1 };
+
+	CU_ASSERT_EQUAL(10, uf->N);
+	CU_ASSERT_EQUAL(6, uf->num_clusters);
+	CU_ASSERT_EQUAL(0, memcmp(uf->clusters, expected_final_ids, 10*sizeof(uint32_t)));
+	CU_ASSERT_EQUAL(0, memcmp(uf->cluster_sizes, expected_final_sizes, 10*sizeof(uint32_t)));
+}
```
Comment (pramsey) [inline liblwgeom/cunit/cu_unionfind.c]: Also put UF_destroy(uf); here

File: liblwgeom/cunit/cu_unionfind.c [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+	uint32_t final_sizes[]    = { 3, 0, 4, 0, 1, 1, 1, 0, 0, 0 };
+	uf->clusters = final_clusters;
+	uf->cluster_sizes = final_sizes;
+	uf->num_clusters = 5;
+
+	uint32_t* actual_ids_by_cluster = UF_ordered_by_cluster(uf);
+	uint32_t expected_ids_by_cluster[] = { 0, 7, 8, 1, 2, 3, 9, 4, 5, 6 };
+
+	CU_ASSERT_EQUAL(0, memcmp(actual_ids_by_cluster, expected_ids_by_cluster, 10 * sizeof(uint32_t)));
+}
```
Comment (pramsey) [inline liblwgeom/cunit/cu_unionfind.c]: Tell me why, when I put a UF_destroy(uf); here, valgrind says I'm doing an invalid free! Scary!
```
==64662== Invalid free() / delete / delete[] / realloc()
==64662==    at 0x1000FD8FD: free (vg_replace_malloc.c:480)
==64662==    by 0x1001357E4: default_freeor (lwutil.c:117)
==64662==    by 0x100136113: lwfree (lwutil.c:216)
==64662==    by 0x10016A6E7: UF_destroy (lwunionfind.c:42)
==64662==    by 0x10003339F: test_unionfind_ordered_by_cluster (cu_unionfind.c:67)
==64662==    by 0x10043C94F: run_single_test (in /opt/local/lib/libcunit.1.dylib)
==64662==    by 0x10043C7AD: CU_run_test (in /opt/local/lib/libcunit.1.dylib)
==64662==    by 0x1000335C3: main (cu_tester.c:202)
==64662==  Address 0x1048f8620 is on thread 1's stack
==64662==  in frame #4, created by test_unionfind_ordered_by_cluster (cu_unionfind.c:54)
==64662== 
==64662== Invalid free() / delete / delete[] / realloc()
==64662==    at 0x1000FD8FD: free (vg_replace_malloc.c:480)
==64662==    by 0x1001357E4: default_freeor (lwutil.c:117)
==64662==    by 0x100136113: lwfree (lwutil.c:216)
==64662==    by 0x10016A6F4: UF_destroy (lwunionfind.c:43)
==64662==    by 0x10003339F: test_unionfind_ordered_by_cluster (cu_unionfind.c:67)
==64662==    by 0x10043C94F: run_single_test (in /opt/local/lib/libcunit.1.dylib)
==64662==    by 0x10043C7AD: CU_run_test (in /opt/local/lib/libcunit.1.dylib)
==64662==    by 0x1000335C3: main (cu_tester.c:202)
==64662==  Address 0x1048f85f0 is on thread 1's stack
==64662==  in frame #4, created by test_unionfind_ordered_by_cluster (cu_unionfind.c:54)
```

**Comment by dbaston** [comment]
@pramsey The "invalid free" is because this test has replaced the `uf->clusters` that was created during `UF_create()` with our own `final_clusters` that was not allocated with `lwalloc`.  The reason for doing this was to test `UF_ordered_by_cluster` independently of the UF algorithm itself.

**Comment by pramsey** [comment]
Is there a nice way to clean it up manually? Getting the liblwgeom tests to valgrind clean is one of our goals (so it's easier to see when we screw up).

**Comment by dbaston** [comment]
Yes, I can add some manual cleanup.

**Comment by pramsey** [comment]
At this point I've cleaned out almost all other leaks (only lwgeom_tcpa ones remain). There are quite a few from your code, and I can't seem them at the test level, so they are probably in the core, where they very much need to be fixed. 

https://pastebin.mozilla.org/8838033

**Comment by dbaston** [comment]
@pramsey I was working through these right now - they call come from
intermediate objects used by the testing code, or from not cleaning up the
results generated by the core.  I'll push a new version shortly.

**Comment by pramsey** [comment]
Good news! I just realized I hadn't looked at all your tests and so most of those could be coming from the test level. Sorry about the alarm.

**Comment by dbaston** [comment]
@pramsey this is what's left.....are these ones OK to stand?
https://pastebin.mozilla.org/8838036

**Comment by pramsey** [comment]
The lexer one should disappear with my most recent updates... the GEOS ones hopefully we can make go away, as the other GEOS stuff doesn't suffer from it.

**Comment by dbaston** [comment]
It looks like it could be related to https://trac.osgeo.org/geos/ticket/587

I don't think any of the other cunit tests are calling initGEOS.

**Comment by pramsey** [comment]
Push changes, looks like we are close enough.

**Comment by dbaston** [comment]
Pushed.  I haven't yet done anything about the `UF_ordered_by_cluster` issue you raised; I'll need to give that some thought.

**Comment by pramsey** [comment]
I'm committing this in a couple minutes, w/ that one test commented out. Gotta beat the feature freeze.

**Comment by dbaston** [comment]
OK, I'm happy to supply a fixed version of the test later on.

**Comment by dbaston** [comment]
@pramsey pushed an updated test; apologies if you got this already.

**Comment by pramsey** [comment]
OK, I'm closing out this RP, as it's all integrated now

**Pull request closed without merging**